gulpfile.js

```diff
@@ -1,5 +1,7 @@
 const gulp = require('gulp');
+const shell = require('gulp-shell');
 const clangFormat = require('gulp-clang-format');
+const os = require('os');
 
```

Comment:
I don't like this approach much. In fact we already removed building stuff from `gulpfile.js` to `package.json` "scripts" time ago and I don't want to depend on `gulp` for the basic building process.

Take into account that this PR also makes "scripts" in `package.json` run `gulp` instead of make, so `gulp` should now be a "dependency" rather than a "devDependency" (otherwise `npm install mediasoup --production` will fail).

Can we figure out a different approach rather than depending on `gulp` for building on Windows? I really want that when executing `npm install mediasoup`, the "postinstall" script calls `make -C worker/` directly (at least in Linux/Mac). Can we do something in the "postinstall" script to make it do something different when in Windows? (without depending on `gulp` I mean).

Reply:
agree. I'll have a try.

Reply:
whether https://www.npmjs.com/package/run-script-os is a better solution ?

Reply:
**run-script-os** looks indeed a perfect solution :)

NOTE: It must be added into "dependencies" since it's required for "postinstall".

Reply:
oops, **run-script-os** seems ok, but can't get `process.env.MEDIASOUP_BUILDTYPE ` which used for judging to build a Debug or Release version.  I'll still work on it.

Reply:
Oh, may be a bug in run-script-os (it overrides env variables)?

Reply:
No. By looking at the [source code](https://github.com/charlesguse/run-script-os/blob/master/index.js) I see no reason for  env variables not to be bypassed. It calls `spawn()` without specifying `env` so, according to the [doc](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options), `process.env` should be used.

Reply:
`process.env.MEDIASOUP_BUILDTYPE` may exist or not. If not, "Release" is assumed. Otherwise it may be set to "Debug".

Regarding run-script-os, I've created the following simple use case and it seems that env variables are properly propagated:

```json
{
  "name": "foo",
  "version": "1.0.0",
  "dependencies": {
    "run-script-os": "^1.0.7"
  },
  "scripts": {
  	"foo": "echo \"--- FOO:${FOO}\"",
  	"test": "run-script-os",
  	"test:darwin": "echo \"--- darwin FOO:${FOO}\"",
  	"test:linux": "echo \"--- linux FOO:${FOO}\""
  }
}
```

and it works fine:

```bash
$ FOO=1234 npm test

> kk@1.0.0 test /private/tmp/kk
> run-script-os


> kk@1.0.0 test:darwin /private/tmp/kk
> echo "--- darwin FOO:${FOO}"

--- darwin FOO:1234
```

```bash
$ FOO=1234 npm run foo

> kk@1.0.0 foo /private/tmp/kk
> echo "--- FOO:${FOO}"

--- FOO:1234
```

Reply:
already updated. but had to use a fork version, this version solves an [issue](https://github.com/charlesguse/run-script-os/issues/20) which caused error after run 'npm install' in windows platform. have better solution?

Reply:
Oh, sad. Ok, honestly I don't want that mediasoup depends on GIT. I've commented in the issue you referenced asking the author for a new version. Not sure if this will happen, so here a different proposal:

What about if we have a `worker/scripts/npm-script-os.js` (a single file hopefully) with the fixed code of that lib (mentioning the author, of course) and use it in "scripts"?:

```json
  "postinstall": "$NODE ./worker/scripts/npm-script-os.js",
  "postinstall:win32": "./worker/win-tasks.sh make",
  "postinstall:default": "make -C worker"
```

NOTE: Such a `./worker/scripts/npm-script-os.js` script should be called with `$NODE` instead of just `node`. As the [doc](https://docs.npmjs.com/cli/run-script) says:

> npm run sets the NODE environment variable to the node executable with which npm is executed.

Reply:
Hi guys! I was kinda just lurking around a little bit and found you're having issues with run-script-os. I had the same issue, so I made the fork.

There's also an npm package under my profile so you can add the dependency based on that...
https://www.npmjs.com/package/@dr.amaton/run-script-os

Good luck!

Reply:
Thanks @kawatzaki, that's great!

Reply:
I'm afraid that the same issue happens (or at least one of the issues I told about above). Basically calling `npm run xxxxxx` within a "scripts" item that uses `npm-script-os` does not work (it does nothing):

```json
"scripts": {
    "test": "npm run test:node && npm run test:worker",
    "test:node": "npm run postinstall && jest",
    "test:worker": "run-script-os",
    "test:worker:win32": "echo test:worker not support yet!!! run test in visual studio!!!",
    "test:worker:default": "make test -C worker",
    "postinstall": "run-script-os",
    "postinstall:win32": "python ./worker/scripts/configure.py --format=msvs && IF %MEDIASOUP_BUILDTYPE% == Release ( MSBuild ./worker/mediasoup-worker.sln /p:Configuration=Release -t:mediasoup-worker ) ELSE ( MSBuild ./worker/mediasoup-worker.sln /p:Configuration=Debug -t:mediasoup-worker )",
    "postinstall:default": "make -C worker"
},
```

Here, if I call `npm test` or `npm run test` this happens:

```
ibc at ibc-macbook in /tmp/mediasoup-win on v3*
$ npm test

> mediasoup@3.2.5 test /private/tmp/mediasoup-win
> npm run test:node && npm run test:worker

> mediasoup@3.2.5 test:node /private/tmp/mediasoup-win
> npm run postinstall && jest


> mediasoup@3.2.5 postinstall /private/tmp/mediasoup-win
> run-script-os

run-script-os was unable to execute the script 'undefined'
```

As you can see, when it calls `npm run postinstall` it does nothing. It is supposed to run `npm-script-os` so it should run the "postinstall:default" script, but it does not.

Reply:
Thanks @kawatzaki.

However I'm afraid that the same issue happens (or at least one of the issues I told about above). Basically calling `npm run xxxxxx` within a "scripts" item that uses `npm-script-os` does not work (it does nothing):

```json
"scripts": {
    "test": "npm run test:node && npm run test:worker",
    "test:node": "npm run postinstall && jest",
    "test:worker": "run-script-os",
    "test:worker:win32": "echo test:worker not support yet!!! run test in visual studio!!!",
    "test:worker:default": "make test -C worker",
    "postinstall": "run-script-os",
    "postinstall:win32": "python ./worker/scripts/configure.py --format=msvs && IF %MEDIASOUP_BUILDTYPE% == Release ( MSBuild ./worker/mediasoup-worker.sln /p:Configuration=Release -t:mediasoup-worker ) ELSE ( MSBuild ./worker/mediasoup-worker.sln /p:Configuration=Debug -t:mediasoup-worker )",
    "postinstall:default": "make -C worker"
},
```

Here, if I call `npm test` or `npm run test` this happens:

```
ibc at ibc-macbook in /tmp/mediasoup-win on v3*
$ npm test

> mediasoup@3.2.5 test /private/tmp/mediasoup-win
> npm run test:node && npm run test:worker

> mediasoup@3.2.5 test:node /private/tmp/mediasoup-win
> npm run postinstall && jest

> mediasoup@3.2.5 postinstall /private/tmp/mediasoup-win
> run-script-os

run-script-os was unable to execute the script 'undefined'
```

As you can see, when it calls `npm run postinstall` it does nothing. It is supposed to run `npm-script-os` so it should run the "postinstall:default" script, but it does not.

Reply:
Oh I see. Ok, let me check on that and I'll see if I can reproduce the issue. I'll add it to my fork repo, be sure to add it to your watch list.

Reply:
Thanks!

Reply:
How to track your repo? Your NPM package does not specify any GitHub (or similar) repository.

Reply:
I'm sorry about that, yeah, here's the repo:

https://github.com/kawatzaki/run-script-os

Anyway, I think the issue is that the 'default' script was not being handled, I fixed it and will be updating the npm package soon enough

Reply:
yup, npm package update to version 1.1.2 please let me know if that fixed it up, if it didn't you can make an issue on my repo and we'll continue from there!

Reply:
Confirmed that it works now, @kawatzaki. Thanks!

Reply:
No problemo, glad I could help!

Reply:
Confirmed, it works. @kawatzaki 👍